docs/invocation.rst

```diff
@@ -42,21 +42,179 @@ Documentation also may be accessed from the command line with
+~~~~~~~~~~~~~~~~~
+
+The RADICAL Pilot remote software components are based in a Python virtual
+environment determined by parameters in the :ref:`RP resource` definition.
+
+By default, RADICAL Pilot resources are configured to bootstrap the target
+environment by creating a fresh virtual environment.
+(``virtenv_mode=create`` and ``rp_version=local`` in most
+`resource <https://radicalpilot.readthedocs.io/en/stable/machconf.html>`__
+definitions.)
+``virtenv_mode=update`` is a better choice than ``create``, so that later
+sessions can re-use a previously bootstrapped pilot venv.
+
+.. _static pilot venv:
+
+Static Pilot venv
+"""""""""""""""""
+
+To minimize the amount of bootstrapping RP performs for each
+:py:class:`~radical.pilot.Session`,
+you can set up a completely static set of virtual environments with customized
+resource definitions in :file:`$HOME/.radical/pilot/configs/`.
+Configure the :ref:`RP resource` to *use* an existing *virtenv* and the RP
+installation it contains.
+Set ``virtenv_mode=use``, ``virtenv=/path/to/venv``, ``rp_version=installed``
+in the RP resource definition.
+
+.. note::
+    This optimization is relevant even for the ``local.localhost`` resource
+    and ``local`` access scheme!
+
+The user (or client) is then responsible for maintaining venv(s) with the
+correct RCT stack (matching the API used by the client-side RCT stack).
+Optionally, the same static venv *may* be used for task execution (see below),
+in which case the user must also maintain a compatible `scalems` installation,
+and any other dependencies of the workflow.
+
+Task environment
+~~~~~~~~~~~~~~~~
+
+shell command injection
+"""""""""""""""""""""""
+
+RP TaskDescriptions allow environment preparation with lines of shell commands
+using :py:attr:`~radical.pilot.TaskDescription.pre_exec`,
+:py:attr:`~radical.pilot.TaskDescription.pre_launch`,
+and :py:attr:`~radical.pilot.TaskDescription.pre_rank`.
+(Note that, in addition to the attribute descriptions,
+RP docs include further discussion at the bottom of the
+:py:class:`~radical.pilot.TaskDescription` class documentation section.)
+
+.. seealso:: https://github.com/SCALE-MS/scale-ms/issues/203
+    for discussion on whether/how to expose this through `scalems.radical`.
+
+static Task venv
+""""""""""""""""
+
+Use :option:`--venv <scalems.radical --venv>` to specify the virtual environment
+in which tasks should execute at the target `resource`.
+The user is responsible for ensuring a compatible `scalems` installation in the
+target venv, as well as for satisfying any other workflow software dependencies.
+
+The RADICAL stack will be made available to the Task (and Master and Worker)
+Python interpreters from the Pilot sandbox through :envvar:`PYTHONPATH`.
+To reproduce the environment seen by your Tasks when interactively using the
+static venv, be sure to include the RP installation in PYTHONPATH.
+
+.. todo:: Clarify the relevant RP installation path and how to discover it.
```

Comment:
We don't currently have a set-up that uses a RP-generated Pilot venv, but static Task venv, so I'm not sure, here.
I think that `$HOME/radical.pilot.sandbox/rp.session.XXX/pilot.YYY/rp_install` may only exist when `prepare_env()` is used. Is this correct?
If `named_env` is used for the Worker with an existing venv instead of `prepare_env()`, what RP installation path is injected for the task?
If the RP installation for the Task is not available until run time, I'm not sure how a user can prepare the static Task venv without a redundant installation of RCT in the venv. Thoughts?

Reply:
* `rp_install`: the Pilot's private RCT stack (which will _not_ be automatically installed in the venv from *create* or *update*). It is injected into the Pilot agent process environment through PYTHONPATH.
* The `prepare_env` task itself will not see the Pilot venv.
* Task environment is initialized with the default environment for the launch method.
* For raptor Tasks, there is not a guarantee that the RCT stack is available: the Task should use a *named_env* or *pre_exec* that explicitly includes access to the RCT stack.

Reply:
@andre-merzky can we try one more time to nail this down? Can you please confirm or correct the following?

If the resource definition specifies *create* or *update*, then the bootstrapper creates a venv and *separately* installs the RCT stack in a `rp_install` directory, which is *not* accessible by default when activating the created venv.

For Pilot agent processes, the `rp_install` directory is injected into PYTHONPATH.

If the resource definition has *use* mode instead of *create* or *update*, `rp_install` is *not* created and not inserted into PYTHONPATH. In that case, the user is responsible for supplying the RCT stack, and can be confident that the Pilot agents are using it.

If Tasks reference the venv, they are not expected to see `rp_install`.

Raptor users are required to make the RCT stack available to the Worker task(s), if it is not available in the default environment of the target resource.

Either `named_env` or `pre_exec` can be provided in the worker description for this purpose.

Question: If the resource specifies *create* or *update*, and the user wants to use the same venv for the Worker process, how should the RCT stack be provided?

Reply:
> If the resource definition specifies _create_ or _update_, then the bootstrapper creates a venv and _separately_ installs the RCT stack in a `rp_install` directory, which is _not_ accessible by default when activating the created venv.

correct. rp_install is created but not inserted e.g. via PYTHON_PATH.

> For Pilot agent processes, the `rp_install` directory is injected into PYTHONPATH.

correct.

> If the resource definition has _use_ mode instead of _create_ or _update_, `rp_install` is _not_ created and not inserted into PYTHONPATH. In that case, the user is responsible for supplying the RCT stack, and can be confident that the Pilot agents are using it.

correct

> If Tasks reference the venv, they are not expected to see `rp_install`.

correct

> Raptor users are required to make the RCT stack available to the Worker task(s), if it is not available in the default environment of the target resource.

correct

> Either `named_env` or `pre_exec` can be provided in the worker description for this purpose.

correct (at least it is intended). Note that `named_env` and `pre_exec` are not mutually exclusive.

> Question: If the resource specifies _create_ or _update_, and the user wants to use the same venv for the Worker process, how should the RCT stack be provided?

This isn't really an intended use case.

scalems + raptor probably needs to rely on a prepared static venv for the Worker (and tasks). Relying on the *update* mode Pilot agent venv is probably not the best way to do this, and using the same venv for Pilot agent and Worker tasks is not a coupling that we should encourage or rely on.

Maybe open an issue to discuss options or features for keeping the Worker venv RCT stack consistent with the client and Pilot agent environments.

In the short term, scalems can update the Worker venv with the already-staged RCT packages at the start of the session. Note that *prepare_env* won't work for that, because it is expected to take no action if the venv already exists. But we could submit a regular command-line task for this before letting the Master launch any Workers.